Add a note about the indices conversion checks in RangePolicy (#558)
kokkos/kokkos-core-wiki: docs/source/API/core/policies/RangePolicy.rst

   Default Constructor uninitialized policy.

> .. cppkokkos:function:: RangePolicy(index_type begin, index_type end)

Reviewer: We need to add a `versionchanged` directive: https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-versionchanged

Author: Thanks @nmm0 ! I was looking for a directive for noting a version just like that one.
But as for this line, it's a bit misleading since the argument types for this constructor were never really `int64_t`. They were templated `member_type` before which were defaulted to `int64_t`, if the policy wasn't constructed with any specified index type. So, the arg type hasn't changed, and this might be more of a correction in the docs.

Reviewer: Ok makes sense, I think it's not needed then

   Provide a begin and end index.

.. cppkokkos:function:: RangePolicy(index_type begin, index_type end, ChunkSize chunk_size)

   Provide a begin and end index as well as a ``ChunkSize``.

.. cppkokkos:function:: RangePolicy(const ExecutionSpace& space, index_type begin, index_type end)

   Provide a begin and end index and an ``ExecutionSpace`` instance to use as the execution resource.

.. cppkokkos:function:: RangePolicy(const ExecutionSpace& space, index_type begin, index_type end, ChunkSize chunk_size)

   Provide a begin and end index and an ``ExecutionSpace`` instance to use as the execution resource, as well as a ``ChunkSize``.

Preconditions:
^^^^^^^^^^^^^^

* The begin index must not be greater than the end index.

> * The actual constructors are templated so we can check that they are converted to ``index_type`` safely (see `#6754 <https://github.com/kokkos/kokkos/pull/6754>`_).
>    * The conversion safety check is performed only if ``index_type`` is convertible to the begin and end index types.

Reviewer 2: ```suggestion
* The actual constructors are templated so we can check that they are converted to ``index_type`` safely if possible (see `#6754 <https://github.com/kokkos/kokkos/pull/6754>`_).
```

Line 144 doesn't work, for the following reasons:

- You have to mention the template types if you want to describe the mechanism of converting to `index_type`.  Saying that `index_type` is convertible to the begin index type, which is `index_type`, doesn't make sense (or course a type is convertible to itself, but that isn't what you are trying to say).
- The conversion check cannot be performed for a type that isn't a fundamental type or a type Kokkos knows about.  For a UDT, even if you can construct it with an `index_type`, you don't know that it has the same value it started with.

I think it is better to just say we make a best effort at checking the conversion.

Author: I agree that it might just be better for us to state that we try to make best effort on this. There weren't really good ways to summarize the general behavior, especially if UDTs are involved, in which case the conversion checks had to be completely skipped.

* The start index must not be greater than the end index.
* The actual constructors are templated so we can check that they are converted to ``index_type`` safely (see `#6754 <https://github.com/kokkos/kokkos/pull/6754>`_).

>    * The conversion safety check is only performed if ``index_type`` is convertible to the start and end index types.

Reviewer 3: start -> begin for consistency

Author: Looks like there were already mixed usages of start/end and begin/end throughout this document.
Changed all `start` to `begin`.  6cd5e99
